File: day09/main.py
from typing import List, Tuple
import collections


M: List[List[int]] = []
for line in open('data.in'):
    row: List[int] = []
    line = line.strip()
    for c in line:
        row.append(int(c))
    M.append(row)

def is_low_point(M: List[List[int]], row: int, col: int) -> bool:
    directions: List[List[int]] = [[1, 0], [-1, 0], [0, 1], [0, -1]]
    for dr, dc in directions:
        r, c = row + dr, col + dc
        if r in range(len(M)) and c in range (len(M[0])) and M[r][c] <= M[row][col]:
            return False
    return True

low_points: List[Tuple[int, int]] = []
for r in range(len(M)):
    for c in range(len(M[r])):
        if is_low_point(M, r, c):
            low_points.append((r, c))


# part I:
risk_level: int = 0
for lp in low_points:
    risk_level += M[lp[0]][lp[1]] + 1

print(risk_level)

# part II:

visit = set()
def bfs(M: List[List[int]], row: int, col: int) -> None:
    q = collections.deque()
    visit.add((row, col))
    q.append((row, col))
    basin = []
    basin.append(M[row][col])
    while q:
        R, C = q.popleft()
        directions: List[List[int]] = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        for dr, dc in directions:
            r, c = R + dr, C + dc
            if r in range(len(M)) and c in range(len(M[0])) and M[r][c] != 9 and (r, c) not in visit:
                q.append((r, c))
                basin.append(M[r][c])
                visit.add((r, c))
    return basin

basins = []
for r in range(len(M)):
    for c in range(len(M[r])):
        if is_low_point(M, r, c) and (r, c) not in visit:
            basins.append(bfs(M, r, c))

basins.sort(key=len, reverse=True)
print(len(basins[0])*len(basins[1])*len(basins[2]))

File: day09/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func SolveI() {
	f, _ := os.Open("input.txt")
	scanner := bufio.NewScanner(f)

	M := [][]int{}
	for scanner.Scan() {
		line := scanner.Text()
		row := []int{}
		for _, c := range line {
			d, _ := strconv.Atoi(string(c))
			row = append(row, d)
		}
		M = append(M, row)
	}

	lowPoints := []int{}
	for i := 0; i < len(M); i++ {
		for j := 0; j < len(M[i]); j++ {
			if isLowPoint(i, j, M) {
				lowPoints = append(lowPoints, M[i][j])
			}
		}
	}
	fmt.Println(riskLevel(lowPoints))

}

func riskLevel(lowPoints []int) int {
	rl := 0
	for _, lp := range lowPoints {
		rl += lp + 1
	}
	return rl
}

func isLowPoint(i, j int, M [][]int) bool {
	var (
		prevX int
		nextX int
		prevY int
		nextY int
	)
	if i-1 < 0 {
		prevX = M[i][j] + 1
	} else {
		prevX = M[i-1][j]
	}

	if i+1 >= len(M) {
		nextX = M[i][j] + 1
	} else {
		nextX = M[i+1][j]
	}

	if j-1 < 0 {
		prevY = M[i][j] + 1
	} else {
		prevY = M[i][j-1]
	}

	if j+1 >= len(M[i]) {
		nextY = M[i][j] + 1
	} else {
		nextY = M[i][j+1]
	}
	point := M[i][j]
	// fmt.Println("testing ", point)
	// fmt.Println("comparing with ", prevX, nextX, prevY, nextY)
	// fmt.Println(prevX > point && nextX > point && prevY > point && nextY > point)
	return prevX > point && nextX > point && prevY > point && nextY > point
}

func main() {
	SolveI()
}
